### PR TITLE
PoC of building disk images inside non privileged containers

### LINKED ISSUE #218: Test and boot the PoC img in AWS
This card is about validating the viability of using the images built in [image_builder](/rancher-sandbox/cOS-toolkit/tree/image_builder) branch and fine tune it. The goal is get up to a point were we can ssh to the recovery system. Documenting the whole process is also part of this card (uploading the image to AWS and booting it with certain cloud-init config).

Finalize the deployment of by re-partitioning and deploying certain cOS like image is out of the scope of this card.

### PR DESCRIPTION
**Pull request by davidcassany**
This is proof of concept for a simple image builder that can be
executed inside a non privileged container.

Fixes #218 

Signed-off-by: David Cassany <dcassany@suse.com>

### CODE REVIEW

**Comment by davidcassany** [comment]
I tested the image with the following procedure:

1. Upload the image to an S3 bucket
```
aws s3 cp <cos-image> s3://cos-images
```

2. Created the disk container JSON (`container.json` file) as:

```
{
  "Description": "cOS Testing image in RAW format",
  "Format": "raw",
  "UserBucket": {
    "S3Bucket": "cos-images",
    "S3Key": "<cos-image>"
  }
}
```

3. Import the disk as snapshot

```
aws ec2 import-snapshot --description "cOS PoC" --disk-container file://container.json
```

4. Followed the procedure described in [AWS docs](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/creating-an-ami-ebs.html#creating-launching-ami-from-snapshot) to register an AMI from snapshot. Used all default settings unless for the firmware, were I forced it to UEFI boot.

5. Launch instance with this simple userdata:
```
name: "Default ec2 user"
stages:
   network.after:
     - name: "Setup user"
       users:
         ec2user:
           name: "ec2user"
           passwd: "$6$M5bXBW/.7pspU1n7$d3Un967.AG8yf9YK2qlUFJt/3EQR57Vrlhtil866FglGY9dI2/arcBCcbSk7/faSq8pkwf1dkD.tDX7iXLOuG1"
           primary_group: "users"
           shell: "/bin/bash"
           homedir: "/home/ec2user"
```

TODOS:

* It is not obvious to create a ec2 user on the fly and authorize it with the SSH key provided by amazon, by default only `root` user gets the key.
* Monitoring features are missing, AWS status checks are not integrated yet.

**Comment by davidcassany** [comment]
Note tests for this PR will fail as it roughly changes the grub config to embrace an squashfs recovery image. This PR should be easy to rebase on top of #192 outcome, then it will be transparent to tests.

File: packages/cloud-config/oem/07_cloud-metadata.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Default cOS OEM configuration file
+#
+# This file is part of cOS and will get reset during upgrades.
+#
+# Before you change this file manually,
+# consider copying this file to /usr/local/cloud-config or
+# copy the file with a prefix starting by 90, e.g. /oem/91_custom.yaml
+name: "Cloud providers metadata"
```
Comment (mudler) [inline packages/cloud-config/oem/07_cloud-metadata.yaml:8]: I'm not sure if we want to embed this directly :thinking:  would it behave _weirdly looking_ if booted into a baremetal?

maybe can go into a `feature, or just all commented if the purpose is to have it as a sample

File: packages/grub-config/config/rec_bootargs.cfg [hunk 2]
```diff
@@ -0,0 +1,3 @@
+set kernel=/boot/vmlinuz
```
Comment (mudler) [inline packages/grub-config/config/rec_bootargs.cfg:1]: I have already those set of changes in my branch for the squashfs change :+1: I think it's safe to merge this after the squashfs change is done, there are also tests to adjust and scripts (which I'm almost done with) to make everything work with the new format

File: packages/cloud-config/oem/07_cloud-metadata.yaml (hunk 1 above)
Comment (davidcassany) [reply packages/cloud-config/oem/07_cloud-metadata.yaml:8]: It does nothing if the provider is not found. Note this plugin is designed to be used against all supported cloud providers without having to touch it. It basically tests all listed providers one after the other until the list is over or until there is one provider that actually succeeds.

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -358,3 +358,37 @@ jobs:
         env:
           GHR_PATH: release/
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+  raw-img:
```
Comment (Itxaka) [inline .github/workflows/build.yaml:362]: can we have a Makefile target to build this locally?

File: images/img-builder.sh [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+# Create root-tree for COS_RECOVERY
+mkdir -p squashfs
+# `luet install` is the only step requiring root
+luet install --system-target squashfs -y system/cos
+mkdir -p root/grub2/x86_64-efi
+mkdir -p root/cOS
+cp squashfs/usr/share/grub2/x86_64-efi/*.mod root/grub2/x86_64-efi
+cp squashfs/etc/cos/grub.cfg root/grub2
+
+mksquashfs squashfs root/cOS/recovery.squashfs -noappend -b 1M -comp xz -Xbcj x86
```
Comment (mudler) [inline images/img-builder.sh:47]: now this one can be consumed directly from the tree e.g. with: `luet install --system-target root/cOS/ -y recovery/cos-squash`

File: .github/workflows/build.yaml (hunk 3 above)
Comment (mudler) [reply .github/workflows/build.yaml:362]: good point! 

giving also another thought - the whole script could even be a package itself, like we do for generating the squashfs directly. The pro would be that we keep all the dependencies in the tree, so the user doesn't need any special tool to generate it. But that's just brain food for a follow-up if we think it makes sense

File: images/img-builder.sh [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    enable: true
+    urls:
+      - build
+    type: disk
+EOF
+
+# Create root-tree for COS_RECOVERY
+mkdir -p squashfs
+# `luet install` is the only step requiring root
+luet install --system-target squashfs -y system/cos
```
Comment (mudler) [inline images/img-builder.sh:41]: mmh, for consistency, would be nice to read those directly from the `manifest.yaml` file, so we can actually specify all of them in one place

File: images/img-builder.sh [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+mkdir -p root/grub2/x86_64-efi
+mkdir -p root/cOS
+cp squashfs/usr/share/grub2/x86_64-efi/*.mod root/grub2/x86_64-efi
```
Comment (mudler) [inline images/img-builder.sh:44]: can we make a package out of this? so we can specify it in the manifest as well for consistency

Comment (davidcassany) [reply images/img-builder.sh]: I wound do it in a follow up issue. I believe the resulting package should also obsolete the `live` bootloader one, so we have a common ground, plus making a cross distro bootloader package might turn into an slightly complex task. I'd vote for a follow up issue.

Comment (davidcassany) [reply images/img-builder.sh]: Opened #279 for that purpose

File: .github/workflows/build.yaml (hunk 3 above)
Comment (davidcassany) [reply .github/workflows/build.yaml:368]: target for the makefile created

**Review by mudler — APPROVED** [review]

**Comment by mudler** [comment]
@davidcassany I think we can merge it also if tests are not passing, there are no any notable change that could actually break them

**Comment by davidcassany** [comment]
I agree, however I see there is a glitch un raw disk job, locally works but on GHA seams to be problematic

**Pull request merged**